appserver/web/war-util/src/main/java/org/glassfish/web/loader/ResourceEntry.java

```diff
@@ -38,40 +48,117 @@ public class ResourceEntry {
 
 
     /**
-     * Binary content of the resource.
+     * Binary content of the resource. May be null.
      */
-    public byte[] binaryContent = null;
+    public byte[] binaryContent;
 
 
     /**
-     * Loaded class.
+     * Loaded class. May be null if the entry is not a class or the class was not loaded.
      */
-    public volatile Class<?> loadedClass = null;
+    public volatile Class<?> loadedClass;
 
 
     /**
-     * URL source from where the object was loaded.
+     * {@link URL} of the entry source.
+     * Can be file, or an url linking inside the JAR file ie. <code>jar:file:/x.jar!/hi.jpg</code>
+     * <p>
+     * For classes it is null after the class is loaded.
      */
-    public URL source = null;
+    public URL source;
 
 
     /**
-     * URL of the codebase from where the object was loaded.
+     * {@link URL} of the entry codebase, ie: <code>file:/x.jar</code>
+     * <p>
+     * For classes it is null after the class is loaded.
      */
-    public URL codeBase = null;
+    public URL codeBase;
 
 
     /**
      * Manifest (if the resource was loaded from a JAR).
+     * <p>
+     * For classes it is null after the class is loaded.
      */
-    public Manifest manifest = null;
+    public Manifest manifest;
 
 
     /**
      * Certificates (if the resource was loaded from a JAR).
+     * <p>
+     * For classes it is null after the class is loaded.
+     */
+    public Certificate[] certificates;
+
+
+    /**
+     * Use this constructor if codebase and source are the same.
+     *
+     * @param url resource location
+     */
+    ResourceEntry(URL url) {
+        codeBase = url;
+        source = url;
+    }
+
+
+    /**
+     * @param codeBase
+     * @param source {@link URL} of the entry source
+     */
+    ResourceEntry(URL codeBase, URL source) {
+        this.codeBase = codeBase;
+        this.source = source;
```

Comment:
These two props should probably be marked `final`?

Reply:
Nope, this part is ugly (still as in original). See javadocs; some attributes are reset when the class is loaded.
Perhaps would be better to have three types of entries: UnloadedClass, LoadedClass and Resource, but finally I stopped here. It would not resolve much, maybe objects would be bit smaller ... maybe we could separate classes and resources, hmm ... maybe next time :-)

Reply:
Ew.